hips/hip-9999.md

```diff
@@ -0,0 +1,317 @@
+- Go native plugin system https://pkg.go.dev/plugin. It does not support Windows
+
+### 🚫 Non-WASM plugins
+
+1. OS/Arch distribtuion problems
+2. Security concerns of non-sandboxed execution
+
+These include:
+- Go native plugin system
+https://pkg.go.dev/plugin
+    - downsides:
+        - does not support Windows. For this reason alone, we can't use it
+        - also the version of go, architecture, and (some other build options) MUST match between parent and plugin
+        - this means WHEN helm core go version upgrades, none of the community plugins will work until the plugin maintainers upgrade their plugin to match the new helm core Go version
+- Hashcorp's go-plugin package
+https://github.com/hashicorp/go-plugin
+    - this is also what protobuf does (subprocess)
+- embedded go interpreter
+    - traefik does this (github.com/traefik/yaegi)
+- Caddy novel Go plugin approach
+https://caddyserver.com/docs/architecture
+
+### 🚫 Non-Extism WASM plugins
+
+- wazero (directly)
+https://github.com/tetratelabs/wazero
+    - note that [extism/go-sdk](https://github.com/extism/go-sdk) (see Extism under [Reference implementation](#reference-implementation)) leverages wazero in a viable way.
+    - downsides:
+        - all params are uint32. So any strings must be interpreted from that
+or in our case structs must be converted in/out (may need to write a library for this?)
+        - The allocation example uses CGO to allocate memory, uses unsafe package. note we don't need to use tinygo
+- Wadge in wasmCloud
+https://wasmcloud.com/docs/developer/languages/go/components#wadge
+    - Bailey Cosmonic CTO suggested (via Taylor Thomas)
+    - downsides:
+        - relies on CGO
+- Envoy's Wasm approach
+https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/wasm_filter
+- WIT/WASI Preview 2 (WASM IDL)
+    - WASI P2 defines WIT as a IDL for wasm modules components. Greatly simplifying message parsing capabilities. However, as WASIP2 is (very) new, support is quite limited (unsupported in wazero for example).
+
+## Open issues
+
+*Any points that are still being decided/discussed.*
+
+TBD
+
+## Future Work
+
+- Expanding plugin functionality to support additional lifecycle phases in addition to the "good candidates for initial plugins" above.
+- Developing advanced tools for debugging and testing plugins.
+- Investigate the need to create migration tools to help users transition from Helm 3 to the Helm 4 plugin-first architecure (or from Chart API v2 to v3). For prior work see the [helm-2to3 plugin](https://github.com/helm/helm-2to3).
+
+## Acknowledgments
+
+WIP
+
+- Special thanks to the Helm community for their feedback and contributions to this proposal.
+ 
+<!-- We can already say:-->
+- Thanks to Extism and Wazero maintainers for guidance
+
+<!-- Ideally we'll also be able to say:
+- Thanks also to Fermyon, Cosmonic, and Dylibso for guidance and collaboration on Wasm integration in Helm.
```

Comment:
```suggestion
- Thanks also to Fermyon, Cosmonic, and Dylibso for guidance and collaboration on WASM integration in Helm.
```

Reply:
Should be Wasm not WASM, even though the doc already contains lots of the latter. Will fix all of these after I commit your other applicable suggestions.

Reply:
Yup. Wasm is the correct spelling